### ISSUE TITLE
Linux Docker Container to automate library downloads

### ISSUE THREAD

**Issue by sarg3nt** [rmcrackan/Libation#395]
**Is your feature request related to a problem? Please describe.**
No

**Describe the solution you'd like**
I'd love a Linux Docker container version that we can set Environment variables to configure.  Many of us run a NAS at home and most NAS platforms now support Docker as standard for software installation.  

**Describe alternatives you've considered**
None that I can think of.

**Additional context**
I run UnRaid and use AudioBookshelf to host my books that Libation downloads for me, only at the moment it's a very manual process.  I'd like a Docker container we could run that we can configure with ENV vars or a config file and that has a method to initialize a library sync and download any new books on a schedule.  
I'm a DevOps engineer and have a decent amount of experience with .Net from back when I was a backend Software Engineer so I might be able to help with this.  I mainly focus on Linux, Docker and Kubernetes these days but could probably help with doing this if you'd like.  My schedule is a bit rough right now, so no promises on it being fast.

**Comment by rmcrackan** [comment]
Sounds promising. I'm very much not a devops person so I'll need your help. (In fact we could really use another one at work -- looking for a job?) What's the difference between your goal and using the current CLI?

**Comment by pixil98** [comment]
I made a docker image awhile ago using the Linux CLI. I run it in a Kubernetes CronJob once a week to download any new books directly into my Plex directory. I didn't document it since I was the only one using it, but I can put together some documentation and put the source up on github. It's pretty basic for now, but it wouldn't be too hard to wrap it in a script to periodically check for books and keep the image running. It should also be possible to have a default Settings.json file that is configured using environment variables.

For now if you want to try it, it's available at:
https://hub.docker.com/r/pixil/libation

It expects the following volumes to be mounted:
/config -- mount your configuration folder here
/db -- optionally mount your database here, if not provided it will assume it's in /config (I wanted to keep my config as configmaps rather than on persistent storage)

Additionally make sure you mount in the folder that Settings.json is configured to save to.

**Comment by wtanksleyjr** [comment]
How does the configuration work? Do you shut down the container, run the
desktop version to set the config and exit it, and then once you start the
container it'll work? Or does one need to filter out path settings first
(obviously you won't see the same environment Docker sees)?

-Wm

**Comment by pixil98** [comment]
Since mine is a one shot scan/download, you should be able to copy your existing Libation config directory, edit the Settings.json to use Linux style paths, then run the container mounting the new directory in as /config. 

Here is the Settings.json file I use in case it helps.
```
{
  "Books": "/data",
  "InProgress": "/tmp",
  "ShowImportedStats": false,
  "ImportEpisodes": false,
  "DownloadEpisodes": false,
  "AllowLibationFixup": true,
  "DecryptToLossy": false,
  "SplitFilesByChapter": false,
  "BadBook": "Retry",
  "FolderTemplate": "<first author>/<if series-><series>/<series#> - <-if series><title short> [<id>]",
  "FileTemplate": "<title> [<id>]",
  "ChapterFileTemplate": "<title> [<id>] - <ch# 0> - <ch title>",
  "Serilog": {
    "MinimumLevel": "Information",
    "WriteTo": [
      {
        "Name": "File",
        "Args": {
          "path": "/var/log/libation.log",
          "rollingInterval": "Month",
          "outputTemplate": "{Timestamp:yyyy-MM-dd HH:mm:ss.fff zzz} [{Level:u3}] (at {Caller}) {Message:lj}{NewLine}{Exception} {Properties:j}"
        }
      }
    ],
    "Using": [
      "Dinah.Core",
      "Serilog.Exceptions"
    ],
    "Enrich": [
      "WithCaller",
      "WithExceptionDetails"
    ]
  },
  "CreateCueSheet": true,
  "RetainAaxFile": false,
  "StripUnabridged": false,
  "StripAudibleBrandAudio": false,
  "LameTargetBitrate": false,
  "LameDownsampleMono": true,
  "LameBitrate": 64,
  "LameConstantBitrate": false,
  "LameMatchSourceBR": true,
  "LameVBRQuality": 2,
  "AutoScan": false,
  "DownloadCoverArt": true,
  "AutoDownloadEpisodes": false,
  "SavePodcastsToParentFolder": false,
  "ChapterTitleTemplate": "<ch#> - <title short>: <ch title>",
  "ReplacementCharacters": {
    "Replacement": [
      {
        "CharacterToReplace": "\u0000",
        "ReplacementString": "_",
        "Description": "All other invalid characters"
      },
      {
        "CharacterToReplace": "/",
        "ReplacementString": "_",
        "Description": "Forward Slash (Filename Only)"
      },
      {
        "CharacterToReplace": "\\",
        "ReplacementString": "_",
        "Description": "Back Slash (Filename Only)"
      },
      {
        "CharacterToReplace": "\"",
        "ReplacementString": "'",
        "Description": "Open Quote"
      },
      {
        "CharacterToReplace": "\"",
        "ReplacementString": "'",
        "Description": "Close Quote"
      },
      {
        "CharacterToReplace": "\"",
        "ReplacementString": "'",
        "Description": "Other Quote"
      },
      {
        "CharacterToReplace": "<",
        "ReplacementString": "{",
        "Description": "Open Angle Bracket"
      },
      {
        "CharacterToReplace": ">",
        "ReplacementString": "}",
        "Description": "Close Angle Bracket"
      },
      {
        "CharacterToReplace": ":",
        "ReplacementString": "-",
        "Description": "Colon"
      },
      {
        "CharacterToReplace": "‘",
        "ReplacementString": "'",
        "Description": "opening single quote"
      },
      {
        "CharacterToReplace": "’",
        "ReplacementString": "'",
        "Description": "closing single quote"
      }
    ]
  },
  "MergeOpeningAndEndCredits": false,
  "BetaOptIn": false
}
```

**Comment by sarg3nt** [comment]
> Since mine is a one shot scan/download, you should be able to copy your existing Libation config directory, edit the Settings.json to use Linux style paths, then run the container mounting the new directory in as /config.
> 
> Here is the Settings.json file I use in case it helps.
> 
> ```
> {
>   "Books": "/data",
>   "InProgress": "/tmp",
>   "ShowImportedStats": false,
>   "ImportEpisodes": false,
>   "DownloadEpisodes": false,
>   "AllowLibationFixup": true,
>   "DecryptToLossy": false,
>   "SplitFilesByChapter": false,
>   "BadBook": "Retry",
>   "FolderTemplate": "<first author>/<if series-><series>/<series#> - <-if series><title short> [<id>]",
>   "FileTemplate": "<title> [<id>]",
>   "ChapterFileTemplate": "<title> [<id>] - <ch# 0> - <ch title>",
>   "Serilog": {
>     "MinimumLevel": "Information",
>     "WriteTo": [
>       {
>         "Name": "File",
>         "Args": {
>           "path": "/var/log/libation.log",
>           "rollingInterval": "Month",
>           "outputTemplate": "{Timestamp:yyyy-MM-dd HH:mm:ss.fff zzz} [{Level:u3}] (at {Caller}) {Message:lj}{NewLine}{Exception} {Properties:j}"
>         }
>       }
>     ],
>     "Using": [
>       "Dinah.Core",
>       "Serilog.Exceptions"
>     ],
>     "Enrich": [
>       "WithCaller",
>       "WithExceptionDetails"
>     ]
>   },
>   "CreateCueSheet": true,
>   "RetainAaxFile": false,
>   "StripUnabridged": false,
>   "StripAudibleBrandAudio": false,
>   "LameTargetBitrate": false,
>   "LameDownsampleMono": true,
>   "LameBitrate": 64,
>   "LameConstantBitrate": false,
>   "LameMatchSourceBR": true,
>   "LameVBRQuality": 2,
>   "AutoScan": false,
>   "DownloadCoverArt": true,
>   "AutoDownloadEpisodes": false,
>   "SavePodcastsToParentFolder": false,
>   "ChapterTitleTemplate": "<ch#> - <title short>: <ch title>",
>   "ReplacementCharacters": {
>     "Replacement": [
>       {
>         "CharacterToReplace": "\u0000",
>         "ReplacementString": "_",
>         "Description": "All other invalid characters"
>       },
>       {
>         "CharacterToReplace": "/",
>         "ReplacementString": "_",
>         "Description": "Forward Slash (Filename Only)"
>       },
>       {
>         "CharacterToReplace": "\\",
>         "ReplacementString": "_",
>         "Description": "Back Slash (Filename Only)"
>       },
>       {
>         "CharacterToReplace": "\"",
>         "ReplacementString": "'",
>         "Description": "Open Quote"
>       },
>       {
>         "CharacterToReplace": "\"",
>         "ReplacementString": "'",
>         "Description": "Close Quote"
>       },
>       {
>         "CharacterToReplace": "\"",
>         "ReplacementString": "'",
>         "Description": "Other Quote"
>       },
>       {
>         "CharacterToReplace": "<",
>         "ReplacementString": "{",
>         "Description": "Open Angle Bracket"
>       },
>       {
>         "CharacterToReplace": ">",
>         "ReplacementString": "}",
>         "Description": "Close Angle Bracket"
>       },
>       {
>         "CharacterToReplace": ":",
>         "ReplacementString": "-",
>         "Description": "Colon"
>       },
>       {
>         "CharacterToReplace": "‘",
>         "ReplacementString": "'",
>         "Description": "opening single quote"
>       },
>       {
>         "CharacterToReplace": "’",
>         "ReplacementString": "'",
>         "Description": "closing single quote"
>       }
>     ]
>   },
>   "MergeOpeningAndEndCredits": false,
>   "BetaOptIn": false
> }
> ```

If you wouldn't mind getting us your Dockerfile I can get it put together with some docs and added to the repo here (if @rmcrackan  doesn't mind)

**Comment by sarg3nt** [comment]
@rmcrackan Sorry, not looking for a job at the moment.  Thanks for asking though!
To answer your question, it's mainly about automation and portability.  An always running Docker container that fires off a loop at the given interval is a lot easier to run and maintain.  When a new docker container comes out Unraid grabs the latest for me.  Sometimes this breaks stuff but it usually doesn't  it also auto backs up the data directories I'm passing in.  Plus some others I'm sure I'm not thinking about.

**Comment by pixil98** [comment]
@sarg3nt It's available at https://github.com/pixil98/docker-libation. I threw together a small readme for it. I can add some more features to it if there's consensus around how an official Libation docker image should function. Right now it's pretty tailored to my needs.

**Comment by rmcrackan** [comment]
> (if @rmcrackan doesn't mind)

rmcrackan is stoked -- I love this kind of stuff when community comes together like this. :smiley: How about you folks who are better than me at docker/devops hash these details out. Let me know when you're done and I'll happily link to it in my [advanced](https://github.com/rmcrackan/Libation/blob/master/Documentation/Advanced.md) documentation.

**Comment by sarg3nt** [comment]
@rmcrackan there are two different ways we can go about this.  The easiest (for us) way would be to continue the work that @pixil98 has already done and use the build artifacts you are already creating.  This is less automated and not the "normal" method for automating Docker container builds.  The more standard way would be to build the code in the Dockerfile.  The advantage of this is that it's easier to automated in a CI/CD pipeline as you don't have external dependencies that aren't checked into source control (i.e. the build artifacts) and we don't run into order of operations issues.  I.e. the Docker build has to happen only after the normal build process completes because otherwise it will fail as the new build artifacts don't exist.  Or they need to be in one build pipeline with on completion steps . . . etc. etc.  I think you get my drift.

And @pixil98, no disrespect at all to the work you've already done, I'm going to get it up and running tonight locally so I have something that works.  Thanks for doing what you have, appreciate it tons!

The below file is from Microsoft's "[Tutorial: Containerize a .NET app](https://learn.microsoft.com/en-us/dotnet/core/docker/build-container?tabs=linux)" page and represents their current boilerplate on building a .net app inside a Dockerfile to create a Docker container.  

```Dockerfile
FROM mcr.microsoft.com/dotnet/sdk:6.0 AS build-env
WORKDIR /App

# Copy everything
COPY . ./
# Restore as distinct layers
RUN dotnet restore
# Build and publish a release
RUN dotnet publish -c Release -o out

# Build runtime image
FROM mcr.microsoft.com/dotnet/aspnet:6.0
WORKDIR /App
COPY --from=build-env /App/out .
ENTRYPOINT ["dotnet", "DotNet.Docker.dll"]
```

I'm 100% fine with doing the majority of the work but would need you for a few things.  

1. The resultant Docker image should go in your repo on Dockerhub (if you don't have one you should create one).  
2. I don't know what your build pipeline looks like.  I don't use Github much because our company works in critical infrastructure and we host everything on prem (which kinda sucks, but that's beside the point)  So, I'm pretty much a noob with any of the cool features of Github
3. What branch and merge strategy do you want to use?  a) Pull repo, branch, push branch and do PR OR b) Fork the repo, do work and PR from fork?   I think the first I need some permissions to push to your repo where the second I don't?  Again, Github noob.  :)

My suggestion would be for the two of us to schedule a call via Discord, Zoom, Teams, whatever floats your boat so you can give me a tour of your code, how your build pipeline works, and we can get some basics out of the way.  Things will go much faster if do.

Thanks!

**Comment by rmcrackan** [comment]
> your repo on Dockerhub (if you don't have one you should create one)

I don't have one. Not even sure what this is. Again: I'm the 'dev' side of devops

> I don't know what your build pipeline looks like.

There isn't one -- it's not automated *at all*. Check-ins do not trigger any CI/CD. No yaml. I have unit tests which I try to manually run frequently. When it's time for a new version, I wrote a separate program to take care of those details. Truly, I'm dev and not ops.

> I don't use Github much

I currently use it for git/versioning, documentation/READMEs, and this issue/ticketing system. At work we use azure devops and we have a dedicated devops team so I rarely have to do more than edit the occasional yml file.

> What branch and merge strategy do you want to use?

I'm mostly a 1 person operation. Since your files and mine won't overlap, just a PR will do. Here's my github cheatsheet (for C# code):

* login to github
* fork repo
* In Visual Studio:
    * clone to desktop (launch VS => Clone a repository => Clone)
    * create new branch
    * make change
    * commit, push
    * `create a pull request` launches github. Must be logged into github in main browser, not incognito
* github:
    * Create pull request

get notified of changes: `https://github.com/{user}/{repo}/commits/master.atom`  
eg: https://github.com/rmcrackan/Libation/commits/master.atom

keep up to date:

* go to the github page for my forked repo
* click button at top: fetch upstream > Fetch and merge

> Discord, Zoom, Teams

Since I really don't have a pipeline, we can probably skip this step.

> What timezone you in?

EST. I'm in NC -- NYC's time zone

**Comment by sarg3nt** [comment]
@rmcrackan 
I'd like to pursue the building of the code in the Dockerfile branch.  What all do I need to build to get that to work? I've got a copy of your repo and am looking at the different projects, there's a lot of them, do I build all of them and copy all of the build artifacts to the target container?  Anything else special needing done?

On a somewhat related note: How often can one rescan the library and liberate, in a loop for example?  i.e. does Audible / Amazon have a limit on API calls per hour / day / whatever or can we do it as much as we want?  
I'm running a 5 minute loop right now to see what it does but was hoping you might know.

**Comment by sarg3nt** [comment]
@rmcrackan 
Also, I'm getting a bunch of these from the CLI
```
DownloadDecryptBook Begin: 05/15/2020 [B088G1N76W] Ep. 4: Aliens!
DownloadDecryptBook Completed: 05/15/2020 [B088G1N76W] Ep. 4: Aliens!
Error processing book. Skipping. This book will be tried again on next attempt. For options of skipping or marking as error, retry with main Libation app.. See log for more details.
```
every time `scan` and `liberate` are called.  I have `  "BadBook": "Ignore"` set but it still does it.  Normal?  Those were free books that are no longer free I can't seem to get rid of them.  Not a big deal, just a bunch of garbage in the output.

**Comment by pixil98** [comment]
No offense taken. I took the approach I did because it was easiest for my use case. If I was integrating it into the main repo I'd use the second approach you outlined.

To that end, I think it would make this easier and @rmcrackan's life easier to set up proper pipelines for this project. Take a look at my fork https://github.com/pixil98/Libation for an example of basic pipelines. For PRs and pushes to main it will automatically build and run tests. On any tag that starts with `v` it will automatically build the project, zip it up, and create a new release. I'm not convinced it's actually building it correctly, and it's definitely only building the windows version, but that could be fixed with a little better understanding of the exact commands you're running to build the various versions.

**Comment by rmcrackan** [comment]
@sarg3nt

> building of the code

Day to day, I just use Visual Studio. When I build a release I have a program which runs a lot of dotnet calls in a certain order. It's big so I'll put it at the bottom.

> How often can one rescan the library and liberate

The Windows UI has a setting to auto-scan which runs every 5 or 6 minutes. I haven't heard of anyone having issues with it. All of this is reverse engineered so there's no way to know the limit for certain.

> free books that are no longer free I can't seem to get rid of them

Yeah I don't have a great solution right now. Libation treats your audible library as the source of truth. Things to try

* Remove them from audible
* In the UI import \> remove can help clean up Libation from things which have been removed from audible
* Settings \> Download/Decrypt tab, top radio buttons will help you skip these
* You can set the download status those books to 'downloaded' to have them not get retried

@pixil98

RE yaml: take a look at the below commands and see if this changes anything for you. Also, I just updated from .net6 => 7

---

Build steps. In order: UI, OS specific module, CLI, optional debugging tool

Windows Classic
```
dotnet clean -c Release
dotnet publish -c Release -o bin\Publish\win-classic LibationWinForms\LibationWinForms.csproj -p:PublishProfile=LibationWinForms\Properties\PublishProfiles\WindowsProfile.pubxml
dotnet publish -c Release -o bin\Publish\win-classic LoadByOS\WindowsConfigApp\WindowsConfigApp.csproj -p:PublishProfile=LoadByOS\WindowsConfigApp\Properties\PublishProfiles\WindowsProfile.pubxml
dotnet publish -c Release -o bin\Publish\win-classic LibationCli\LibationCli.csproj -p:PublishProfile=LibationCli\Properties\PublishProfiles\WindowsProfile.pubxml
dotnet publish -c Release -o bin\Publish\win-classic HangoverWinForms\HangoverWinForms.csproj -p:PublishProfile=HangoverWinForms\Properties\PublishProfiles\WindowsProfile.pubxml
```
Windows Chardonnay
```
dotnet clean -c Release
dotnet publish -c Release -o bin\Publish\win-chardonnay LibationAvalonia\LibationAvalonia.csproj -p:PublishProfile=LibationAvalonia\Properties\PublishProfiles\WindowsProfile.pubxml
dotnet publish -c Release -o bin\Publish\win-chardonnay LoadByOS\WindowsConfigApp\WindowsConfigApp.csproj -p:PublishProfile=LoadByOS\WindowsConfigApp\Properties\PublishProfiles\WindowsProfile.pubxml
dotnet publish -c Release -o bin\Publish\win-chardonnay LibationCli\LibationCli.csproj -p:PublishProfile=LibationCli\Properties\PublishProfiles\WindowsProfile.pubxml
dotnet publish -c Release -o bin\Publish\win-chardonnay HangoverAvalonia\HangoverAvalonia.csproj -p:PublishProfile=HangoverAvalonia\Properties\PublishProfiles\WindowsProfile.pubxml
```
Linux Chardonnay
```
dotnet clean -c Release
dotnet publish -c Release -o bin\Publish\linux-chardonnay LibationAvalonia\LibationAvalonia.csproj -p:PublishProfile=LibationAvalonia\Properties\PublishProfiles\LinuxProfile.pubxml
dotnet publish -c Release -o bin\Publish\linux-chardonnay LoadByOS\LinuxConfigApp\LinuxConfigApp.csproj -p:PublishProfile=LoadByOS\LinuxConfigApp\Properties\PublishProfiles\LinuxProfile.pubxml
dotnet publish -c Release -o bin\Publish\linux-chardonnay LibationCli\LibationCli.csproj -p:PublishProfile=LibationCli\Properties\PublishProfiles\LinuxProfile.pubxml
dotnet publish -c Release -o bin\Publish\linux-chardonnay HangoverAvalonia\HangoverAvalonia.csproj -p:PublishProfile=HangoverAvalonia\Properties\PublishProfiles\LinuxProfile.pubxml
```
Mac Chardonnay
```
dotnet clean -c Release
dotnet publish -c Release -o bin\Publish\macos-chardonnay LibationAvalonia\LibationAvalonia.csproj -p:PublishProfile=LibationAvalonia\Properties\PublishProfiles\MacOSProfile.pubxml
dotnet publish -c Release -o bin\Publish\macos-chardonnay LoadByOS\MacOSConfigApp\MacOSConfigApp.csproj -p:PublishProfile=LoadByOS\MacOSConfigApp\Properties\PublishProfiles\MacOSProfile.pubxml
dotnet publish -c Release -o bin\Publish\macos-chardonnay LibationCli\LibationCli.csproj -p:PublishProfile=LibationCli\Properties\PublishProfiles\MacOSProfile.pubxml
dotnet publish -c Release -o bin\Publish\macos-chardonnay HangoverAvalonia\HangoverAvalonia.csproj -p:PublishProfile=HangoverAvalonia\Properties\PublishProfiles\MacOSProfile.pubxml
```

**Comment by pixil98** [comment]
@rmcrackan I have a much simpler setup for the pipeline now, it builds all 4 versions with a single workflow. I'm still struggling to get them to build correctly, especially the Linux and MacOS versions. I don't have much experience with C#, but one thing I noticed is that, on line three of your scripts, it's referencing a non existent pubxml file for the profile when building the main Libation project. The actual profiles are in the `LoadByOS\<os>ConfigApp\PublishProfiles` folders rather than `LoadByOS\<os>ConfigApp\Properties\PublishProfiles`.

**Comment by pixil98** [comment]
@rmcrackan I think I have it all working now. Can you take a look at this [release ](https://github.com/pixil98/Libation/releases/tag/v1.0.12) and let me know if the builds all look correct? If it looks good I can make a pull request to merge the pipelines into your repo. There will be a little bit of fine tuning to do to fit the release process you want, but we can do that in the pull request.

@sarg3nt Once the pipelines merge in, I think the easiest way to build a docker image would be to use the Linux build artifact from earlier in the pipeline. That way the docker image and the released zip will be the exact same build and there isn't a chance of them differing

**Comment by sarg3nt** [comment]
@pixil98 Great work!  I did get the Dockerfile built to do a direct build of the files but I totally agree that if it's practical to use the build artifacts from a previous step then we should.  
What's your plan moving forward.  I was just basing what I was doing off of the Dockerfile work you had already done.  If you want to continue the pipeline work to build the docker images I'm good with that.  
I do have a more advanced `liberate.sh` file I can contribute.  I'll just paste it here as it's probably the only bits you'd need that I've done.

@rmcrackan Couple questions for you.  
1. Are all the projects you list for the Linux build necessary for a command line only implementation of Libation?  i.e. in a Docker container could we remove any of the output (UI for example) and have it all still work?  It's a bit of a micro-optimization but the full directory is 181 MB and smaller Docker files are always better.  :)
2. I don't think that `"Badbook": "Ignore"` is working in the Linux Command line as it tries those broken files every loop.
3. It's been a while since I wrote .net code but I seem to remember that the config provider had a way of loading config from either a `settings.json` or environment variables.  Do you know if you support the environment vars?  I haven't tested it yet and figured it was easier to ask.  I'm asking because right now setting up the Docker container is a bit cumbersome.  Right now we are  copying in all the entire config directory.  I notice that libation command line will create the config directory if it does not exist, at that point all the user would need to copy in is their existing database and `AccountsSettings.json` if I'm not mistaken (please correct me where I'm wrong), the common thing to do with Docker containers then is to set any of the settings they want to be different from the defaults via ENV vars.  i.e if all I want to change is `BadBook` then I could pass `BadBook=Ignore` as an env var to the container.  One important note, dashes work for Windows but do not work with Linux env vars, underscores only (learned that lesson the hard way), or Camel case which is what you are currently doing is fine too (technically, public env vars in Linux are supposed to be Upper snake case ME_BIG_ENV_VAR, but it's really more of a strongly worded suggestion than a technical requirement and I wouldn't worry about it too much, though if you have to implement each one by hand, might as well make them the standard.  I kind of remember some way to just tell the configuration system to enable environment variable support and it will work, but that was web apps, and several years ago.  If env vars are a no go then would a partial `Settings.json` file work?  Looking through the code I think I see where it is setting defaults for any missing options in `Settings.json`?  Let us know which you would like to go with, ENV vars or partial `Settings.json` or as is.  Are there any other bits of the Libation config folder that is necessary to get it to work?  Right now I'm just copying everything in and it's working, just trying to figure out how to simplify the directions and setup for new users. 
4. I think I saw you mention somewhere in the docs that the command line version won't necessarily do upgrades.  Is that ever going to change?  I probably won't touch the GUI again for a while but would like the Docker instance I have looping to upgrade the DB properly.  If not, we probably need some way to inform the user and make sure nothing breaks when a new version of the Docker container spins up automatically.  
5. Any plans to do a Linux ARM build?  I just pre-ordered some Orange Pi 5s for a Kubernetes cluster.  Not a big deal as I can keep running this on my NAS, just curios.  :)
6.  I can't find it in the docs now but I thought I saw somewhere where you say that `[<id>]` needs to be in part of the path or the file as Libation uses it for tracking?  Am I remembering that wrong?  What does that mean?  I was under the impression that Libation kind of saves the files and doesn't really look at them again, it just goes by it's DB as to what it has liberated.  Do we need the `[<id>]` in the directory or file name?
7. It would be super useful if we could remove a book from the DB using the command line, thus causing the command line to re-add it and download it again.  I've needed this a couple times for various reasons.  

BTW, I don't know if I mentioned this, but my primary use case is to get my Audible books into [audiobookshelf](https://www.audiobookshelf.org/) which is a self hosted Audiobook (and soon ebook) system with an amazing UI and great phone apps.  I host my own copy and listen to my books off my own servers now.  Super cool.  Mainly love it because Audibles self discover of what I own is really terrible.  i.e. I have a couple hundred books I haven't read but it's really hard to slice and dice them in Audible but amazing in Audiobookshelf, mainly the series support is outstanding.  Just thought I'd share.  
At some point I'll add to the docs section here how to set up Libation to create Audiobookshelf compatible directory structures.
It's pretty much mainly just this:  `"FolderTemplate": "<author>/<if series-><series>/<series#> - <-if series><title>",`

@pixil98 here's the updated `liberate.sh` file

```bash
#!/bin/bash

# Rewire echo to print date time
echo() {
    if [[ ! -z $1  ]]; then
        printf "`date`: $1\n"
    else
        printf ""
    fi
}

# ################################
# Setup
# ################################
echo "****************************************************************"
echo "Beginning startup"
echo "****************************************************************"
if [[ -z "${SLEEP_TIME}" ]]; then
  SLEEP_TIME_LOCAL="30m"
  echo "No sleep time passed in via the SLEEP_TIME env var, setting a default of 30 minutes"
else
  SLEEP_TIME_LOCAL="${SLEEP_TIME}"
  echo "Sleep time is set to ${SLEEP_TIME_LOCAL}"
fi

echo ""

# Check if the config directory is passed in, and there is no link to it then create the link.
if [ -d "/config" ] && [ ! -d "/root/Libation" ]; then
    echo "Linking config directory to the Libation config directory"
    ln -s /config/ /root/Libation
fi

# If no config error and exit
if [ ! -d "/config" ]; then
    echo "    ERROR: No /config directory.  You must pass in a volume containing your config mapped to /config"
    exit 1
fi

# If user passes in db from a /db/ folder and a db does not already exist / is not already linked
FILE=/db/LibationContext.db
if [ -f "${FILE}" ] && [ ! -f "/config/LibationContext.db" ]; then
    echo "Linking passed in Libation database from /db/ to the Libation config directory"
    ln -s $FILE /config/LibationContext.db
fi

# Confirm we have a db in the config direcotry.
if [ ! -f "/config/LibationContext.db" ]; then
    echo "    ERROR: No Libation database detected, exiting."
    exit 2
fi

# ################################
# Loop and liberate
# ################################
while true
do
    echo ""
    echo ""
    echo "**** Starting scan"
    /libation/LibationCli scan
    echo "**** Starting Liberation"
    /libation/LibationCli liberate
    echo ""
    echo "**** Waiting ${SLEEP_TIME_LOCAL} to liberate again."
    sleep ${SLEEP_TIME_LOCAL}
done
```
And here's what a run looks like in the container.
```
Fri Nov 18 22:44:03 PST 2022: ****************************************************************
Fri Nov 18 22:44:03 PST 2022: Beginning startup
Fri Nov 18 22:44:03 PST 2022: ****************************************************************
Fri Nov 18 22:44:03 PST 2022: Sleep time is set to 5m
Fri Nov 18 22:44:03 PST 2022: Linking config directory to the Libation config directory
Fri Nov 18 22:44:03 PST 2022: **** Starting scan
Scanning Audible library. This may take a few minutes.
Scan complete.
Total processed: 666
New: 0
Fri Nov 18 22:44:28 PST 2022: **** Starting Liberation
DownloadDecryptBook Begin: 05/15/2020 [B088G1N76W] Ep. 4: Aliens!
DownloadDecryptBook Completed: 05/15/2020 [B088G1N76W] Ep. 4: Aliens!
Error processing book. Skipping. This book will be tried again on next attempt. For options of skipping or marking as error, retry with main Libation app.. See log for more details.
DownloadDecryptBook Begin: 05/15/2020 [B088G39G8G] Ep. 3: Robots & Artificial Intelligence
DownloadDecryptBook Completed: 05/15/2020 [B088G39G8G] Ep. 3: Robots & Artificial Intelligence
Error processing book. Skipping. This book will be tried again on next attempt. For options of skipping or marking as error, retry with main Libation app.. See log for more details.
DownloadDecryptBook Begin: 05/15/2020 [B088GSYJMP] Ep. 2: Elevating Black Voices
DownloadDecryptBook Completed: 05/15/2020 [B088GSYJMP] Ep. 2: Elevating Black Voices
Error processing book. Skipping. This book will be tried again on next attempt. For options of skipping or marking as error, retry with main Libation app.. See log for more details.
DownloadDecryptBook Begin: 01/09/2021 [B00B77MXYQ] Stranded
DownloadDecryptBook Completed: 01/09/2021 [B00B77MXYQ] Stranded
Error processing book. Skipping. This book will be tried again on next attempt. For options of skipping or marking as error, retry with main Libation app.. See log for more details.
Done. All books have been processed
Fri Nov 18 22:44:51 PST 2022: **** Waiting 5m to liberate again.
```

**Comment by rmcrackan** [comment]
Thanks for your patience. Busy weekend.

@pixil98

> referencing a non existent pubxml file

D'oh! Fixed. They're all nested in the Properies\\PublishProfiles hierarchy now. Jeez I wonder how many problems that's caused me this year without realizing it. Thanks! When those were created there was a lot of churn and re-org. Must have missed this part.

> Can you take a look at this release

Aside from the zip file names not matching the version, the Windows ones seem good.

@sarg3nt

1) Are all the projects you list for the Linux build necessary for a command line only implementation of Libation?

tl;dr

No, for Linux CLI you only need to build that one.

longer:

LoadByOS loads OS specific stuff shared by UI, CLI, and the Hangover debug tool. In practice, so far I'm only using this for windows UI and CLI. It's there (and empty) for linux and mac for ease of use later. Those other 3 (UI, CLI, Hangover) are unrelated programs which are packaged together for ease. There are no interdependencies yet. I forsee a future where Hangover may one day call the CLI. For now, that's very far off.

2) Please put in the linux beta thread or open a separate bug report

3) I'm not loading anything from Env vars, it's all from files. appsettings.json lives next to the executable and says where to look for all other settings. If that dir exists, then it won't look for a user directory (nor create a new config directory within that user directory). Settings.json has most of the remaining config. Audible accounts and tokens are in AccountSettings.json. FileLocations.json is a convenience cache and does very little. Without a UI, you'll have no use for this. QuickFilters.json is the search settings -- again, you won't need it in docker. These haven't changed in a while but there's no guarantee they won't in the future. These files are implementation details, not apis.

5) I think I saw you mention somewhere in the docs that the command line version won't necessarily do upgrades.

I have no immediate plans to tackle this. The happy path is trivial. The edge cases are too difficult. If you use the CLI to do an upgrade then try to run it again during the upgrade, shit goes sideways fast. It's even more likely if it happens automatically. I think the CLI announces upgrades though so that would be easy to script against. Or just use the github api to check for a newer version. Once you strip away the upgrade itself, that's really all I'm doing behind the scenes.

6) Any plans to do a Linux ARM build?

I'm willing to compile to anything if someone finds the right dotnet publish commands similar to the above

7) I can't find it in the docs now but I thought I saw somewhere where you say that [<id>] needs to be in part of the path or the file as Libation uses it for tracking?

Sorry for the confusion, this gets nuanced. You're mostly right. The one time that Libation looks at files is if you click the green stoplight then Libation will attempt to open the folder where that file lives. As you implied: I make very little effort to do anything with files post decrypt. People commonly download files then move them, which I encourage. If you keep the `ID` in the path, then Libation can locate the book again if somewhere inside the tree of that download directory. If you download and don't move it, then FileLocations.json will remember the orig location. For anything else, Libation can no longer corrolate the file with the database entry. It still remembers that it was downloaded, it just don't know where it was moved to.

audiobookshelf sounds impressive. How does it compare to self-hosting Plex?

> At some point I'll add to the docs section

Thanks, that sounds great

**Comment by wtanksleyjr** [comment]
> audiobookshelf sounds impressive. How does it compare to self-hosting
Plex?

I use both; I've tried several times to use Plex for audiobooks and given
up every time (it's great for movies, but absolutely frustrating to use for
audiobooks). First time I set up audiobookshelf it worked for me for my
Android, and within a few months the iOS client app worked well enough for
my sister to use. Unlike Plex, it has built-in editors for metadata, as
well as pre-configured book lookups.

-Wm

**Comment by sarg3nt** [comment]
Hey all, FYI, I'm going to be unavailable for getting any real work done on this for about 3 weeks.  We are going on a cruise on Saturday, then my wife is having surgery the next week and the recovery is going to be a bit long so I'll be out of town.
I should be available for comments / questions / etc. here the week after next though.
Happy Thanksgiving!

**Comment by pixil98** [comment]
I have everything working and generating builds/releases with a naming convention very similar to the current official builds. The only differences are `windows` instead of `win` and the classic windows build matches the same style as the other builds. This is because it vastly simplifies the pipeline if I can use the same OS variable everywhere and have everything build the same way. In preparation to create the pull request I pulled in the changes to the official repo. Now it won't build anymore. It seems like the dotnet 7 publish command is trying to  perform the restore step as dotnet 6 and it's failing. I've been looking into workarounds, but I think it's going to require an update on Microsoft's end. Here is an issue tracking the same problem with azure-pipelines https://github.com/microsoft/azure-pipelines-tasks/issues/17152.

**Comment by rmcrackan** [comment]
@pixil98 Can you give an example of your final zip names?

**Comment by pixil98** [comment]
@rmcrackan you can see a release of the current pull request at https://github.com/pixil98/Libation/releases/tag/v1.0.15.

The names are
```
Libation.1.0.15-linux-chardonnay.zip
Libation.1.0.15-macos-chardonnay.zip
Libation.1.0.15-windows-chardonnay.zip
Libation.1.0.15-windows-classic.zip 
```

**Comment by rmcrackan** [comment]
Unfortunately the classic zip needs to come first alphabetically for backward compatibility reasons. The old updater would get the available assets (eg: Libation.6.7.4.zip , Source code zip , Source code tar.gz) and use the 1st as the application's zip. If anyone is using one of the old versions which checks that way, then they're using classic. This is why in recent releases it annoyingly starts with `Classic-`

**Comment by rmcrackan** [comment]
In recent versions, the updater uses [this regex](https://github.com/rmcrackan/Libation/blob/master/.releaseindex.json) to find the correct zip. Which means I can easily edit `win` to `windows`. Classic still needs to come first though.

**Comment by pixil98** [comment]
I can put in special handling to keep classic at the top. Does the new updater check the regex in the repo before getting the latest version? Otherwise they'll still break if it changes. I've already updated the regex as part of the pull request.

**Comment by rmcrackan** [comment]
> Does the new updater check the regex in the repo before getting the latest version?

Yes. The only annoyance is that the old updater used the 1st zip. The new updater always checks the live json/regex vs the latest release, so they can stay in sync and we're free to change them at will. And if we mess up, deploy a new one and that's the one that will be checked. It's much more flexible now.

**Comment by pixil98** [comment]
Okay, I added a line to the release workflow that specially renames the windows classic build to put Classic in front of it. It's a little unfortunate because it means the release workflow now knows a bit about the various builds instead of keeping all of that contained in the build workflow, but what's the point of technical debt if you don't get to hard code exceptions every once in a awhile.

https://github.com/pixil98/Libation/releases/tag/v1.0.17
```
Classic-Libation.1.0.17-windows-classic.zip
Libation.1.0.17-linux-chardonnay.zip
Libation.1.0.17-macos-chardonnay.zip
Libation.1.0.17-windows-chardonnay.zip 
```

**Comment by rmcrackan** [comment]
I admit: I've lost the plot regarding the original request. What is needed to fulfill this ticket?

**Comment by pixil98** [comment]
It needs a docker build step added to the release pipeline. I have one that's almost done. If you want you can assign this issue to me.

**Comment by pixil98** [comment]
I have a docker image building automatically on the reworked pipelines. Currently it's publishing the images to docker hub at pixil/libation-test. @rmcrackan do you have a docker hub account? I just need to know the name of the image you want published, probably `<your account>/libation`. Your username and an auth token for docker will be setup later as repository secrets.

**Comment by rmcrackan** [comment]
My account is `rmcrackan1` . (I think I set up the rmcrackan account at my last job and I no longer have access to that email. Oops.) Just let me know what to do and I'll be glad to configure it as needed.

**Comment by rmcrackan** [comment]
@pixil98 With the completion of your latest PR: pending a new publish, is this issue closed?

**Comment by rmcrackan** [comment]
@sarg3nt The docker image is available. My account there is `rmcrackan` , repo is `libation`
